modules/net-vpc-factory/README.md

```diff
@@ -0,0 +1,837 @@
+- [Projects](#projects)
+- [VPCs](#vpcs)
+  - [Subnets](#subnets)
+  - [Firewall Rules](#firewall-rules)
+  - [Cloud NAT](#cloud-nat)
+  - [Routes](#routes)
+  - [Private Service Access (PSA)](#private-service-access-psa)
+  - [VPC DNS Policy](#vpc-dns-policy)
+- [Connectivity](#connectivity)
+  - [VPC Peering](#vpc-peering)
+  - [Cloud VPN](#cloud-vpn)
+    - [GCP to OnPrem](#gcp-to-onprem)
+    - [GCP to GCP VPN](#gcp-to-gcp-vpn)
+  - [Network Connectivity Center (NCC)](#network-connectivity-center-ncc)
+    - [NCC VPC Spokes](#ncc-vpc-spokes)
+    - [NCC VPN Spokes](#ncc-vpn-spokes)
+- [DNS](#dns)
+  - [Private Zone](#private-zone)
+  - [Forwarding Zone](#forwarding-zone)
+  - [Peering Zone](#peering-zone)
+  - [Public Zone with DNSSEC](#public-zone-with-dnssec)
+- [Variables](#variables)
+- [Outputs](#outputs)
+- [TODO](#todo)
+<!-- END TOC -->
+
+## Factory configuration
+
+At a high level, the factory consumes all YAML files in a directory (whose path is `var.factories_config.vpcs`).
+Each file defines a single project, which can either be created by the factory or pre-created externally, and all the network infrastructure to be deployed in that project, including
+
+- Project
+  - NCC Hubs
+  - VPCs
+    - Connectivity (including VPNs, VPC peerings, NCC VPC Spokes and NCC VPN hybrid spokes)
+    - DNS policies
+    - DNS zones
+    - Firewall rules/policies (via a sub-factory)
+    - NAT (public or private)
+    - PSA configuration
+    - Routes (including PBRs)
+    - Subnets (via a sub-factory)
+
+### Configuration Methods
+
+- **YAML Factory:** The primary method is to define project and network configurations within individual YAML files placed in the directory specified by `var.factories_config.vpcs`. This factory approach allows for modular and organized management of complex setups.
+- **Direct Variable Input (Discouraged):** The module also defines `var.network_project_config`. While it's *technically possible* to populate this variable directly in a .tfvars file, this is **not the recommended approach**. It bypasses the intended factory pattern and makes managing multiple project configurations cumbersome. The variable definition primarily serves as **documentation** for the structure expected within the YAML configuration files.
+
+### Cross-Referencing Resources
+
+A key capability of this factory is establishing relationships between resources defined across different VPCs or projects. This is achieved through cross-referencing using logical keys within the YAML configuration. These keys strictly follow the pattern `project_key/vpc_key/resource_key` for resources like VPCs, VPN gateways, routers, or NCC hubs/groups. During Terraform execution, the factory code translates these logical keys into the actual resource IDs or self-links required by the underlying modules. This  mechanism avoids hardcoding resource IDs and enables the declarative definition of complex topologies. It's extensively used in configurations for:
```

Comment:
s/This  mechanism/This mechanism/ (extra space)

Reply:
eagle eye